ts/nni_manager/common/trial_keeper/process.ts

```diff
@@ -0,0 +1,238 @@
+import { Logger, getLogger } from 'common/log';
+
+export interface TrialProcessOptions {
+    command: string;
+    shell?: string;
+    codeDirectory: string;
+    outputDirectory: string;
+    commandChannelUrl: string;
+    platform: string;
+    sequenceId?: number;
+    environmentVariables: Record<string, string>;
+}
+
+interface TrialProcessInfo {
+    startSuccess?: boolean;  // it is true when the shell successfully launched, even if the trial command is ill-formed
+    startTime?: number;
+    stopTime?: number;
+    stopCode?: number | null;
+    stopSignal?: string | null;
+}
+
+export class TrialProcess {
+    private id: string;
+    private info: TrialProcessInfo = {};
+    private log: Logger;
+    private proc: ChildProcess | null = null;
+    private started: Deferred<void> = new Deferred();
+    private stopped: Deferred<void> = new Deferred();
+
+    constructor(id: string) {
+        this.id = id;
+        this.log = getLogger(`TrialProcess.${id}`);
+    }
+
+    /**
+     *  Spawn the trial process. Return success or not.
+     *
+     *  Note that the trial is considered "success" here even if the trial command is ill-formed
+     *  or it returns non-zero code immediately.
+     *
+     *  This is because the command is run in a shell and
+     *  this method only checks if the shell is successfully launched.
+     *
+     *  If `options.shell` is empty or not set, it will use "/bin/bash" on UNIX,
+     *  or "ComSpec" environment variable on Windows (default behavior of Node.js).
+     *  When the shell is powershell, the exit code is likely to be 0/1 instead of concrete number.
+     **/
+    public async spawn(options: TrialProcessOptions): Promise<boolean> {
+        // might change for log collection
+        const stdout = fs.createWriteStream(path.join(options.outputDirectory, 'trial.stdout'));
+        const stderr = fs.createWriteStream(path.join(options.outputDirectory, 'trial.stderr'));
+        await Promise.all([ events.once(stdout, 'open'), events.once(stderr, 'open') ]);
+
+        let shell: string | boolean = true;
+        if (options.shell) {
+            shell = options.shell;
+        } else if (process.platform !== 'win32') {
+            shell = '/bin/bash';
+        }
+
+        const spawnOptions: SpawnOptions = {
+            cwd: options.codeDirectory,
+            env: this.buildEnv(options),
+            stdio: [ 'ignore', stdout, stderr ],
+            shell: shell,
+        };
+
+        this.proc = child_process.spawn(options.command, spawnOptions);
+
+        this.proc.on('spawn', () => { this.resolveStart('spawn'); });
+        this.proc.on('exit', (code, signal) => { this.resolveStop('exit', code, signal); });
+        this.proc.on('error', (err) => { this.handleError(err); });
+        this.proc.on('close', (code, signal) => { this.resolveStop('close', code, signal); });
+
+        await this.started.promise;
+        return Boolean(this.info.startSuccess);
```

Comment:
Reminds me that there is no logging here in case of failure.

Reply:
It's logged when setting `startSuccess` to `false`.
At here we don't know why it fails.